core/src/main/java/org/apache/calcite/sql/SqlBinaryOperator.java

```diff
@@ -140,7 +140,13 @@ private RelDataType convertType(SqlValidator validator, SqlCall call, RelDataTyp
     if (SqlTypeUtil.inCharFamily(operandType0)
         && SqlTypeUtil.inCharFamily(operandType1)) {
       Charset cs0 = operandType0.getCharset();
+      if (call.operand(0) instanceof SqlBasicCall) {
```

Comment:
Is this the right place for this check?
Why does the operandType0 have the wrong charset?
Maybe the bug is in the place where the charset for operand0 type was inferred.

Reply:
You are right. I've added inferReturnType for both CONVERT and TRANSLATE function.